parser/grammar.y

```diff
@@ -312,6 +311,15 @@ rule
       }
       condition '}'
       {
+        // Walk the condition AST looking for string identifiers that are not
```

Comment:
This could be done more efficiently without traversing the condition's AST for every rule. We could have a map where we add the identifiers for the strings belonging to the rule currently being parsed. This map would be emptied when we finish parsing a rule, and it would be populated in the `string_declaration` production, just when the corresponding `ast.TextString`, `ast.HexString` or `ast.RegexpString` is being created. The keys are the strings identifiers while the values could be anything, we don't actually need the value, we are using the map as a set. 

Then, every time you are about to create a `ast.StringIdentifier` node, you check the map first, looking for a key that matches the string that is being used in the condition, if not found in the map, the string was not previously defined. This is the approach followed in YARA.

This map must be stored in the `lexer` (just as the global list of rules is stored in `lexer.ruleSet`)

Reply:
I thought about doing it that way but dismissed it for some reason. I'll revisit it today.